src/loggers.ts

```diff
@@ -0,0 +1,95 @@
+import { CancelError, HTTPError, Options, Response } from 'got';
+
+export interface Logger {
+  info(...args: unknown[]): void;
+  error(...args: unknown[]): void;
+  debug(...args: unknown[]): void;
+}
+
+export interface HttpLogger {
+  logSuccess(response: Response, options: Options): void;
+  logFailure(error: HTTPError | CancelError): void;
+}
+
+export class NoOpLogger implements HttpLogger {
+  // eslint-disable-next-line class-methods-use-this
+  logSuccess(_response: Response, _options: Options): void {}
+
+  // eslint-disable-next-line class-methods-use-this
+  logFailure(_error: HTTPError | CancelError): void {}
+}
+
+export class GenericLogger implements HttpLogger {
+  readonly #logger: Logger;
+
+  constructor(logger: Logger) {
+    this.#logger = logger;
+  }
+
+  logSuccess(response: Response, options: Options): void {
+    const { url, statusCode, timings } = response;
+    const message = `${options.method} ${url} ${statusCode} ${new Date().getTime() - timings.start} ms`;
+    if (options.method === 'GET') {
+      this.#logger.debug(message);
+    } else {
+      this.#logger.info(message);
+    }
+    this.#logger.debug('request-options', JSON.stringify(options).replace(/\\n/g, ''));
+  }
+
+  logFailure(error: HTTPError | CancelError): void {
+    const { context, headers, method } = error.options;
+    const { requestUrl } = error.request ?? {};
+    const code = error.response?.statusCode ?? error.code;
+    const { start, end, error: err } = error?.timings ?? {};
+    const duration = err && end && start ? (err ?? end) - start : undefined;
+    this.#logger.error(
+      '\n' +
+        '--------------------------------------------------------------------\n' +
+        `[${method}] ${requestUrl} ${code ?? 'unknown statusCode'} (${duration ?? ' - '} ms)\n` +
+        `headers: ${JSON.stringify(headers)}\n` +
+        `request-options: ${JSON.stringify({ ...error.options, context }).replace(/\\n/g, '')}\n` +
+        `error:${error.message}\n` +
+        `stack:${error.stack}\n` +
+        '--------------------------------------------------------------------',
+    );
+  }
+}
+
+export class PinoLogger implements HttpLogger {
+  readonly #logger: Logger;
+
+  constructor(logger: Logger) {
+    this.#logger = logger;
+  }
+
+  logSuccess(res: Response): void {
+    const { request: req, timings } = res;
+    const logLevelToInvoke = req.options.method === 'GET' ? 'debug' : 'info';
+    this.#logger[logLevelToInvoke]({
+      req,
+      res,
+      responseTime: Number(timings?.end) - Number(timings?.start),
```

Comment:
I need to change this to return `null` or `undefined` since JSON doesn't support `NaN` because JSON is overrated crap.

That code looks so ugly though, so I'm punting it for now.